bdai_ros2_wrappers/bdai_ros2_wrappers/futures.py

```diff
@@ -1,25 +1,27 @@
 # Copyright (c) 2023 Boston Dynamics AI Institute Inc.  All rights reserved.
 from threading import Event
-from typing import Any, Optional
+from typing import Optional
 
+from rclpy.context import Context
 from rclpy.task import Future
+from rclpy.utilities import get_default_context
 
 
-def wait_for_future(future: Future, timeout_sec: Optional[float] = None) -> bool:
+def wait_for_future(future: Future, timeout_sec: Optional[float] = None, *, context: Optional[Context] = None) -> bool:
     """Blocks while waiting for a future to become done
 
     Args:
         future (Future): The future to be waited on
         timeout_sec (Optional[float]): An optional timeout for how long to wait
+        context (Optional[Context]): Current context (will use the default if none is given)
 
     Returns:
         bool: True if successful, False if the timeout was triggered
     """
+    if context is None:
+        context = get_default_context()
     event = Event()
-
-    def done_callback(_: Any) -> None:
-        nonlocal event
-        event.set()
-
-    future.add_done_callback(done_callback)
-    return event.wait(timeout=timeout_sec)
+    context.on_shutdown(event.set)
```

Comment:
What happens if the context has multiple events set and shuts down? Is there a possibility where two events can be waiting on each other and deadlock?

Reply:
If the context has multiple events, all of them will be set. Events do not wait on each other. Callees wait on events. Callees could run into deadlock if, for example, they have to complete each other futures or you somehow manage to wait on future within an on context shutdown handler. 

So I would say deadlock are a non-issue inasmuch as it won't happen by accident unless code is flawed.